### PR TITLE
Add Split String Expression

### PR DESCRIPTION
**Pull request by sheelio**

### CODE REVIEW

File: corehq/apps/userreports/expressions/specs.py [hunk 1]
```diff
@@ -331,3 +331,28 @@ def _get_subcases(self, case_id, context):
...
+        string_value = self._string_expression(item, context)
+        if not isinstance(string_value, basestring):
+            return None
+
+        index_value = self._index_expression(item, context)
+        if not isinstance(index_value, int):
+            return None
+
+        try:
+            return string_value.split(delimiter)[index_value]
```
Comment (dimagimon) [inline corehq/apps/userreports/expressions/specs.py]: F821 undefined name 'delimiter'

File: corehq/apps/userreports/tests/test_expressions.py [hunk 2]
```diff
@@ -1036,3 +1036,70 @@ def test_cache(self):
...
+                "string_expression": {
+                    "type": "property_name",
+                    "property_name": "string_property"
+                },
+                "index_expression": {
+                    "type": "property_name",
+                    "property_name": "index_property"
+                }
+            })
+            self.assertEqual(expected, split_string_expression({"string_property": original, "index_property": index}))
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: F821 undefined name 'original'
E501 line too long (119 > 115 characters)

File: corehq/apps/userreports/tests/test_expressions.py [hunk 3]
```diff
@@ -1036,3 +1036,70 @@ def test_cache(self):
...
+        ]:
+            split_string_expression = ExpressionFactory.from_spec({
+                "type": "split_string",
+                "string_expression": {
+                    "type": "property_name",
+                    "property_name": "string_property"
+                },
+                "index_expression": index
+            })
+            self.assertEqual(expected, split_string_expression({"string_property": original}))
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: F821 undefined name 'original'

File: corehq/apps/userreports/tests/test_expressions.py [hunk 4]
```diff
@@ -1036,3 +1036,70 @@ def test_cache(self):
...
+            split_string_expression = ExpressionFactory.from_spec({
+                "type": "split_string",
+                "string_expression": {
+                    "type": "property_name",
+                    "property_name": "string_property"
+                },
+                "delimiter": ",",
+                "index_expression": index
+            })
+            self.assertEqual(expected, split_string_expression({"string_property": original}))
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: F821 undefined name 'original'

File: corehq/apps/userreports/tests/test_expressions.py [hunk 5]
```diff
@@ -1036,3 +1036,70 @@ def test_cache(self):
...
+                "string_expression": {
+                    "type": "property_name",
+                    "property_name": "string_property"
+                },
+                "index_expression": {
+                    "type": "property_name",
+                    "property_name": "index_property"
+                }
+            })
+            self.assertEqual(expected, split_string_expression({"string_property": string_value, "index_property": index}))
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: E501 line too long (123 > 115 characters)

File: corehq/apps/userreports/tests/test_expressions.py [hunk 6]
```diff
@@ -1036,3 +1036,73 @@ def test_cache(self):
...
+                    "type": "property_name",
+                    "property_name": "string_property"
+                },
+                "index_expression": {
+                    "type": "property_name",
+                    "property_name": "index_property"
+                }
+            })
+            self.assertEqual(expected, split_string_expression({
+                "string_property": string_value,
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: W291 trailing whitespace

**Comment by sheelio** [comment]
Up for anyone to review this and merge.

**Comment by sheelio** [comment]
Pinging interrupt to review and merge
@czue @emord @mkangia

**Comment by benrudolph** [comment]
nice sheel 💪

**Comment by czue** [comment]
:+1: sorry it took so long for me to get to this. one quick thought i had was whether it made sense for the split expression to return a list and then use an `array_index` expression into that list (more generic but also more annoying to write). in the other direction another thought i had was whether we should add indexing and split to the list of allowed functions in `evaluator`. @sheelio did you consider either of those options?

**Comment by sheelio** [comment]
I looked at having split just return a list and then use array_index, but seemed annoying/complicated to use.  Its easy enough to implement that way though too. 

Didn't look into evaluator though (but could if you think it makes sense).

**Comment by czue** [comment]
nope, not advocating for either of those, was just wondering. thanks.

**Pull request merged**